### ISSUE TITLE
pack build should support `--extension`

### ISSUE THREAD

**Issue by natalieparellano** [buildpacks/pack#1551]
### Description
Today you can run `pack build --buildpack` to dynamically add buildpacks to the builder and modify the order. It should be possible to add extensions in the same way.

### Proposed solution
* `--extension`s should always come first
* There is a separate order for extensions (`order-extensions` in order.toml) so we'll need to take that into account
* ~This should be blocked on https://github.com/buildpacks/pack/issues/1470 (support for `pack extension package`) as `--extension` can accept a package input~ - we can make a separate issue for extension packages

### Additional context
- [x] This feature should be documented somewhere

**Comment by edithwuly** [comment]
Hi @natalieparellano, I would like to have a try at this issue.

**Comment by natalieparellano** [comment]
That is awesome @edithwuly! I can assign the issue to you if you like. In case it's helpful, here are a few relevant places in the code:
* https://github.com/buildpacks/pack/blob/0a19002ffa717aa6fe21c90ea58630ce6b14d3e1/pkg/client/build.go#L133 - `Buildpacks` holds additional buildpacks from `--buildpack`; we'll likely want to make an `Extensions`
* https://github.com/buildpacks/pack/blob/0a19002ffa717aa6fe21c90ea58630ce6b14d3e1/internal/commands/build.go#L161 - where we read in the build options
* https://github.com/buildpacks/pack/blob/0a19002ffa717aa6fe21c90ea58630ce6b14d3e1/pkg/client/build.go#L751 - where we download the additional buildpacks; we can likely re-use some of this function for extensions; however the project descriptor part is not relevant and instead of appending to `bldr.Order()`, we want to append to `bldr.OrderExtensions()`

Please don't hesitate to reach out if you have any questions!

**Comment by edithwuly** [comment]
Thanks @natalieparellano! It will be great if you could assign this issue to me. And the places you share are exactly what I'm looking for. I will let you know as soon as there is progress.

**Comment by jjbustamante** [comment]
Thanks a lot for helping us with this @edithwuly!

**Comment by edithwuly** [comment]
Hi @natalieparellano, I have looking into the code and have some idea about what to do. 
- https://github.com/buildpacks/pack/blob/6219ee5a3a912e3a9d1ad75e1c1c65500064f50d/pkg/client/build.go#L133
  - I'm going to create a field `Extensions` to hold additional extensions from `--extension`

- https://github.com/buildpacks/pack/blob/6219ee5a3a912e3a9d1ad75e1c1c65500064f50d/pkg/client/build.go#L751
  - I'm going to create a function `processExtensions()`  to fetch and order extensions. I believe I can re-use some code of `processBuildpacks()`. And I notice there are many conditions considered in `processBuildpacks()` like `locatorType` and `opts.PreBuildpacks`. Is there any condition I need to consider when processing extensions?

- https://github.com/buildpacks/pack/blob/6219ee5a3a912e3a9d1ad75e1c1c65500064f50d/pkg/client/build.go#L942
  - I'm going to pass the fetched extensions and order to `createEphemeralBuilder()`, and add them to the builder by invoking `bldr.AddExtension()` and `bldr.SetOrderExtensions()`

Is there anything I missed?

**Comment by edithwuly** [comment]
@natalieparellano sounds great! Thanks for your help! I'm working on it.

**Comment by edithwuly** [comment]
Hi @natalieparellano! I have completed a simple version and proposed a PR. Please help me review, thanks!

Besides, when developing, I have some questions:
- About the order of extensions. I'm not sure whether the added extensions should be placed at first, and I just re-used `prependBuildpackToOrder()` since there is no pre or post concept for extensions. 
- I handle `RegistryLocator` in `processExtensions()` and re-used `fetchBuildpack()`  to fetch extensions by just adding `ModuleKind` value in `DownloadOptions` structure. Is there any conditions I didn't consider? 
- Should I do some unit tests or end-to-end tests? I have tested this version for tutorial case on Buildpack official website. Maybe you can help me get more test cases!

**Comment by natalieparellano** [comment]
@edithwuly thank you so much for the PR! I'll look this over tomorrow :)
